### ISSUE TITLE
Ignore ellipsis input

### ISSUE THREAD

**Issue by mapidentity** [kkinnear/zprint#188]
Is there an option to ignore ellipsis (...) in input? My REPL output is constrained and too long data get shown as, for example;

```clojure
{:a 1,
 :b 2,
 ...}
```

zprint fails with `Failed to zprint: java.lang.IllegalArgumentException: No value supplied for key: ...`

I think these ellipsis can be treated as whitespace.

**Comment by kkinnear** [comment]
What an interesting idea!   I'm familiar with that form of output:
```
(czprint (get-options) {:max-length 5})
{:agent {:object? false},
 :array {:hex? false, :indent 1, :object? false, :wrap? true},
 :atom {:object? false},
 :binding {:flow? false,
           :force-nl? false,
           :ha-depth-factor 0,
           :ha-width-factor 0,
           :hang-accept nil,
           ...},
 :cache {:directory ".zprint", :location "HOME"},
 ...}
nil
```
I do have two questions.  

1. In the above example, the zprint output is to stdout, and `czprint` returns `nil`.

Is the "thing" that you are feeding to zprint that contains a `...` a string, or is it a structure?  By structure, I mean -- could you do `(first thing)` on it, and get `[:a 1]`? This is a distinction that doesn't come up much, but makes a great deal of difference to how I would implement this in zprint. 

Another way of asking that question is to say:  When you feed thing thing with `...` in it to zprint, do you just say:
```
(zprint thing)
```
or do you say:
```
(zprint thing {:parse-string? true})
```

2. Do you want the `...` to appear in the output from zprint?  I would assume that you would, but wanted to be sure.

Both of these questions are designed to help me better understand the context in which you are using zprint in the REPL.

I'm considering changing the `:max-length` output for zprint when doing maps to `... ...`, for what that is worth.

Thanks!

**Comment by mapidentity** [comment]
1. I'm passing the data through stdin to the command line tool of zprint, so yes, as string input. I wouldn't know how to create this in a datastructure tbh, because the example given is an unbalanced map.

2. yes, I would like to keep the ellipsis in the output for my use case, so that I can still see that there is actually more information than there is displayed.

Thank you for considering this.

**Comment by kkinnear** [comment]
First, I have to say that I said to myself "what what I thinking???" when I read your response.  Of course a map can't be an actual structure with a `...` as a single thing.

That said, I'm still struggling to understand your use case.  It makes perfect sense that you are passing the structure you describe above into the command line version of zprint.  However, when I do that, I don't get the error that you report, I get a map that is not lovely because `...` sorts to the front of the keys, but it doesn't provoke an error.

Using the MacOS command line tools from the last several versions:
```
% more cljfiles/i188.clj
{:a 1,
 :b 2,
 ...}
% ./zprintm-1.1.1 <cljfiles/i188.clj
{..., :a 1, :b 2}
% ./zprintm-1.1.0 <cljfiles/i188.clj
{..., :a 1, :b 2}
% ./zprintm-1.0.0 <cljfiles/i188.clj
{..., :a 1, :b 2}
```
This is probably more what you want.  I could see making the appearance of `...` disable sorting, to make this the default output.
```
% ./zprintm-1.1.1 '{:map {:sort? false}}' <cljfiles/i188.clj
{:a 1, :b 2, ...}
```
But I'm not having any luck provoking the error that you encountered.  

The error `Failed to zprint:` comes from only one place, but it doesn't happen for me. 

How about you try this, and send me the large amount of output:
```
./zprintm-1.1.1 '{:dbg? true}' <yourfile.clj
```
where you replace `./zprintm-1.1.1` with the command line tool you are using, and feed it the file through `stdin` however you do that.  

The only thing you need to add is: `'{:dbg? true}'`.  That will give me lots of information, and probably let me figure out why I can't reproduce this.

You might also include the output from:
```
./zprintm-1.1.1 -e
```
which will show me everything you've changed from the defaults in the options map.  There may be something I'm missing here.  Well, there is certainly something I'm missing here, and it is possibly a result of however zprint in configured in your environment.

Thanks!

**Comment by kkinnear** [comment]
I have found where the error that you received comes from by searching the Clojure source:
```
 (array-map :a 1 :b 2 '...)
Execution error (IllegalArgumentException) at zprint.core/eval4402 (form-init808968820290687725.clj:1).
No value supplied for key: ...
```
but I have no idea why (or how) anything would be trying to create an actual array-map out of the input on `stdin` to the zprint command-line tool.

**Comment by kkinnear** [comment]
I now think I know sort of how this is happening -- the parsing library I'm using will successfully parse an "unbalanced" map such as the one you are dealing with.  You can ask the parser to turn something it has parsed into an "sexpr", and if you do that with this unbalanced map, you do get the error you see:
```
Execution error (IllegalArgumentException) at rewrite-clj.node.seq/map-node$fn (seq.clj:110).
No value supplied for key: ...
```
While I've tried to figure out what in your configuration might be doing this, I'm as yet unable to actually reproduce the direct problem.  This makes the output from `./zprintm-1.1.1 -e` even more important, as it is probably something in your configuration that is different from the default that is triggering this issue.

Thanks for chasing this down with me, as not only does it impact what you want to do, but it also may cause issues for some new capabilities I'm about to release.

**Comment by kkinnear** [comment]
This now works in `1.1.2`:
```
zprint.core=> (print i188)
{:a 1,
 :b 2,
 ...}
nil
zprint.core=> (czprint i188 {:parse-string? true})
{:a 1, :b 2, ...}
nil
zprint.core=> (czprint i188 {:parse-string? true :map {:force-nl? true}})
{:a 1,
 :b 2,
 ...}
nil
```
I hope that it helps.  Making this work was spawned several new configuration capabilities.

**Comment by mapidentity** [comment]
Thanks for supporting this use case! It seems to work for most scenarios. I've just run into one thing: (1.1.2)

```shell
zprintm '{:parse {:ignore-if-parse-fails #{"..."}}}' -e
```

```clojure
{:parse {:ignore-if-parse-fails {:set-by "command-line options",
                                 :value #{"..."}}},
 :search-config? {:set-by "Home directory file: /myhomedir/.zprintrc",
                  :value true}}
```

This works:

```shell
echo '[{:a 1 ...}{:b 2 ...}]' | zprintm '{:parse {:ignore-if-parse-fails #{"..."}}}'
```

Gives:

```clojure
[{:a 1, ...} {:b 2, ...}]
```

This doesn't work (list instead of vector):
```shell
echo '({:a 1 ...}{:b 2 ...})' | zprintm '{:parse {:ignore-if-parse-fails #{"..."}}}'
```

Which gives:
```
Failed to zprint: java.lang.IllegalArgumentException: No value supplied for key: ...
```

Same applies for sets, i.e. `#{{:a 1 ...}{:b 2 ...}}`.

When leaving out the `{:parse {:ignore-if-parse-fails #{"..."}}}` option, result is the same.

Making it a quoted list, i.e. with ' or (quote ...) gives the same results. This matters because my REPL outputs lists results this way.

**Comment by kkinnear** [comment]
Thanks for trying this out.  I managed to pick only the things that worked to test, it seems.  I have tried out of bunch of other things (in addition to what you pointed out), and I think I've got this working now -- having made a lot of changes.  It will be fixed in `1.2.0`.  Though I've been impressed by the number of places where this could go wrong, so we will see.

**Comment by kkinnear** [comment]
I fixed this (again) in `1.2.0`.

**Comment by mapidentity** [comment]
Thank you, this is much appreciated!
